nav2_map_server/src/map_saver.cpp

```diff
@@ -30,221 +31,173 @@
 
 #include "nav2_map_server/map_saver.hpp"
 
-#include <cstdio>
-#include <fstream>
-#include <memory>
-#include <string>
-#include <utility>
-#include <vector>
-
-#include "Magick++.h"
-#include "nav2_map_server/map_mode.hpp"
-#include "nav_msgs/msg/occupancy_grid.h"
-#include "nav_msgs/srv/get_map.hpp"
-#include "rclcpp/rclcpp.hpp"
-#include "tf2/LinearMath/Matrix3x3.h"
-#include "tf2/LinearMath/Quaternion.h"
-#include "yaml-cpp/yaml.h"
+#include <stdexcept>
 
 namespace nav2_map_server
 {
-MapSaver::MapSaver(const rclcpp::NodeOptions & options)
-: Node("map_saver", options), save_next_map_promise{}
+MapSaver::MapSaver()
+: nav2_util::LifecycleNode("map_saver")
 {
-  Magick::InitializeMagick(nullptr);
-  {
-    mapname_ = declare_parameter("output_file_no_ext", "map");
-    if (mapname_ == "map") {
-      mapname_ += "_" + std::to_string(static_cast<int>(now().seconds()));
-    }
-    threshold_occupied_ = declare_parameter("threshold_occupied", 65);
-    if (100 < threshold_occupied_) {
-      throw std::runtime_error("Threshold_occupied must be 100 or less");
-    }
-    threshold_free_ = declare_parameter("threshold_free", 25);
-    if (threshold_free_ < 0) {
-      throw std::runtime_error("Free threshold must be 0 or greater");
-    }
-    if (threshold_occupied_ <= threshold_free_) {
-      throw std::runtime_error("Threshold_free must be smaller than threshold_occupied");
-    }
+  RCLCPP_INFO(get_logger(), "Creating");
 
-    std::string mode_str = declare_parameter("map_mode", "trinary");
-    try {
-      map_mode = map_mode_from_string(mode_str);
-    } catch (std::invalid_argument &) {
-      map_mode = MapMode::Trinary;
-      RCLCPP_WARN(
-        get_logger(), "Map mode parameter not recognized: '%s', using default value (trinary)",
-        mode_str.c_str());
-    }
+  save_map_timeout_ = std::make_shared<rclcpp::Duration>(
+    std::chrono::milliseconds(declare_parameter("save_map_timeout", 2000)));
 
-    image_format = declare_parameter("image_format", map_mode == MapMode::Scale ? "png" : "pgm");
-    std::transform(
-      image_format.begin(), image_format.end(), image_format.begin(),
-      [](unsigned char c) {return std::tolower(c);});
-    const std::vector<std::string> BLESSED_FORMATS{"bmp", "pgm", "png"};
-    if (
-      std::find(BLESSED_FORMATS.begin(), BLESSED_FORMATS.end(), image_format) ==
-      BLESSED_FORMATS.end())
-    {
-      std::stringstream ss;
-      bool first = true;
-      for (auto & format_name : BLESSED_FORMATS) {
-        if (!first) {
-          ss << ", ";
+  free_thresh_default_ = declare_parameter("free_thresh_default", 25),
+  occupied_thresh_default_ = declare_parameter("occupied_thresh_default", 65);
+
+  map_listener_ = rclcpp::Node::make_shared("map_listener");
```

Comment:
You can remove this node and instead use the ` nav2_util::LifecycleNode("map_saver", true)` constructor (note, `true`), that creates inside of `LifecycleNode` a `rclcpp_node_` that will spin up a thread to spin independently and is consistent with how we built the other servers. You can then use `rclcpp_node_` for anything you use `map_listener_` for.

Reply:
Thank you for this advice! This is very useful to know.

Reply:
Fixed starting from https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317R27